core/src/main/java/com/netflix/conductor/core/reconciliation/WorkflowSweeper.java

```diff
@@ -117,12 +117,12 @@ void unack(WorkflowModel workflowModel, long workflowOffsetTimeout) {
         long postponeDurationSeconds = 0;
         for (TaskModel taskModel : workflowModel.getTasks()) {
             if (taskModel.getStatus() == Status.IN_PROGRESS) {
-                if (taskModel.getTaskType().equals(TaskType.TASK_TYPE_WAIT)
-                        || taskModel.getTaskType().equals(TaskType.TASK_TYPE_HUMAN)) {
-                    postponeDurationSeconds =
-                            (taskModel.getWaitTimeout() != 0)
-                                    ? taskModel.getWaitTimeout() + 1
-                                    : workflowOffsetTimeout;
+                if (taskModel.getTaskType().equals(TaskType.TASK_TYPE_WAIT)) {
+                    long deltaInSeconds =
```

Comment:
what happens if the wait task has no timeout? This can happen if the WAIT is configured to wait for an external trigger without a timeout value.

Reply:
I didn't take that case into account because it didn't seem to me that the Wait class allowed it, as it requires the use of duration or until. The only way to have a zero timeout is by setting the until parameter to the start of the epoch, January 1, 1970, at 0 hours.

However, I imagine it might be possible in another way that I'm not aware of.

Would you like me to share a code snippet that takes that case into account?

Reply:
I'm not sure how to update the code. Should I create another Pull Request?

Reply:
`core/src/main/java/com/netflix/conductor/core/reconciliation/WorkflowSweeper.java`
```java
/*
 * Copyright 2022 Netflix, Inc.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */
package com.netflix.conductor.core.reconciliation;

import java.time.Instant;
import java.util.Optional;
import java.util.Random;
import java.util.concurrent.CompletableFuture;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;

import com.netflix.conductor.annotations.VisibleForTesting;
import com.netflix.conductor.common.metadata.tasks.TaskDef;
import com.netflix.conductor.common.metadata.tasks.TaskType;
import com.netflix.conductor.core.WorkflowContext;
import com.netflix.conductor.core.config.ConductorProperties;
import com.netflix.conductor.core.exception.NotFoundException;
import com.netflix.conductor.core.execution.WorkflowExecutor;
import com.netflix.conductor.dao.QueueDAO;
import com.netflix.conductor.metrics.Monitors;
import com.netflix.conductor.model.TaskModel;
import com.netflix.conductor.model.TaskModel.Status;
import com.netflix.conductor.model.WorkflowModel;

import static com.netflix.conductor.core.config.SchedulerConfiguration.SWEEPER_EXECUTOR_NAME;
import static com.netflix.conductor.core.utils.Utils.DECIDER_QUEUE;

@Component
public class WorkflowSweeper {

    private static final Logger LOGGER = LoggerFactory.getLogger(WorkflowSweeper.class);

    private final ConductorProperties properties;
    private final WorkflowExecutor workflowExecutor;
    private final WorkflowRepairService workflowRepairService;
    private final QueueDAO queueDAO;

    private static final String CLASS_NAME = WorkflowSweeper.class.getSimpleName();

    @Autowired
    public WorkflowSweeper(
            WorkflowExecutor workflowExecutor,
            Optional<WorkflowRepairService> workflowRepairService,
            ConductorProperties properties,
            QueueDAO queueDAO) {
        this.properties = properties;
        this.queueDAO = queueDAO;
        this.workflowExecutor = workflowExecutor;
        this.workflowRepairService = workflowRepairService.orElse(null);
        LOGGER.info("WorkflowSweeper initialized.");
    }

    @Async(SWEEPER_EXECUTOR_NAME)
    public CompletableFuture<Void> sweepAsync(String workflowId) {
        sweep(workflowId);
        return CompletableFuture.completedFuture(null);
    }

    public void sweep(String workflowId) {
        WorkflowModel workflow = null;
        try {
            WorkflowContext workflowContext = new WorkflowContext(properties.getAppId());
            WorkflowContext.set(workflowContext);
            LOGGER.debug("Running sweeper for workflow {}", workflowId);

            if (workflowRepairService != null) {
                // Verify and repair tasks in the workflow.
                workflowRepairService.verifyAndRepairWorkflowTasks(workflowId);
            }

            workflow = workflowExecutor.decide(workflowId);
            if (workflow != null && workflow.getStatus().isTerminal()) {
                queueDAO.remove(DECIDER_QUEUE, workflowId);
                return;
            }

        } catch (NotFoundException nfe) {
            queueDAO.remove(DECIDER_QUEUE, workflowId);
            LOGGER.info(
                    "Workflow NOT found for id:{}. Removed it from decider queue", workflowId, nfe);
            return;
        } catch (Exception e) {
            Monitors.error(CLASS_NAME, "sweep");
            LOGGER.error("Error running sweep for " + workflowId, e);
        }
        long workflowOffsetTimeout =
                workflowOffsetWithJitter(properties.getWorkflowOffsetTimeout().getSeconds());
        if (workflow != null) {
            long startTime = Instant.now().toEpochMilli();
            unack(workflow, workflowOffsetTimeout);
            long endTime = Instant.now().toEpochMilli();
            Monitors.recordUnackTime(workflow.getWorkflowName(), endTime - startTime);
        } else {
            LOGGER.warn(
                    "Workflow with {} id can not be found. Attempting to unack using the id",
                    workflowId);
            queueDAO.setUnackTimeout(DECIDER_QUEUE, workflowId, workflowOffsetTimeout * 1000);
        }
    }

    @VisibleForTesting
    void unack(WorkflowModel workflowModel, long workflowOffsetTimeout) {
        long postponeDurationSeconds = 0;
        for (TaskModel taskModel : workflowModel.getTasks()) {
            if (taskModel.getStatus() == Status.IN_PROGRESS) {
                if (taskModel.getTaskType().equals(TaskType.TASK_TYPE_WAIT)) {
                    if (taskModel.getWaitTimeout() == 0) {
                        postponeDurationSeconds =
                                properties.getWorkflowOffsetTimeout().getSeconds();
                    } else {
                        long deltaInSeconds =
                                (taskModel.getWaitTimeout() - System.currentTimeMillis()) / 1000;
                        postponeDurationSeconds = (deltaInSeconds > 0) ? deltaInSeconds + 1 : 1;
                    }
                } else if (taskModel.getTaskType().equals(TaskType.TASK_TYPE_HUMAN)) {
                    postponeDurationSeconds = properties.getWorkflowOffsetTimeout().getSeconds();
                } else {
                    postponeDurationSeconds =
                            (taskModel.getResponseTimeoutSeconds() != 0)
                                    ? taskModel.getResponseTimeoutSeconds() + 1
                                    : workflowOffsetTimeout;
                }
                break;
            } else if (taskModel.getStatus() == Status.SCHEDULED) {
                Optional<TaskDef> taskDefinition = taskModel.getTaskDefinition();
                if (taskDefinition.isPresent()) {
                    TaskDef taskDef = taskDefinition.get();
                    if (taskDef.getPollTimeoutSeconds() != null
                            && taskDef.getPollTimeoutSeconds() != 0) {
                        postponeDurationSeconds = taskDef.getPollTimeoutSeconds() + 1;
                    } else {
                        postponeDurationSeconds =
                                (workflowModel.getWorkflowDefinition().getTimeoutSeconds() != 0)
                                        ? workflowModel.getWorkflowDefinition().getTimeoutSeconds()
                                                + 1
                                        : workflowOffsetTimeout;
                    }
                } else {
                    postponeDurationSeconds =
                            (workflowModel.getWorkflowDefinition().getTimeoutSeconds() != 0)
                                    ? workflowModel.getWorkflowDefinition().getTimeoutSeconds() + 1
                                    : workflowOffsetTimeout;
                }
                break;
            }
        }
        queueDAO.setUnackTimeout(
                DECIDER_QUEUE, workflowModel.getWorkflowId(), postponeDurationSeconds * 1000);
    }

    /**
     * jitter will be +- (1/3) workflowOffsetTimeout for example, if workflowOffsetTimeout is 45
     * seconds, this function returns values between [30-60] seconds
     *
     * @param workflowOffsetTimeout
     * @return
     */
    @VisibleForTesting
    long workflowOffsetWithJitter(long workflowOffsetTimeout) {
        long range = workflowOffsetTimeout / 3;
        long jitter = new Random().nextInt((int) (2 * range + 1)) - range;
        return workflowOffsetTimeout + jitter;
    }
}
```

Reply:
`core/src/test/java/com/netflix/conductor/core/reconciliation/TestWorkflowSweeper.java`

```java
/*
 * Copyright 2022 Netflix, Inc.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */
package com.netflix.conductor.core.reconciliation;

import java.time.Duration;
import java.util.List;
import java.util.Optional;

import org.junit.Before;
import org.junit.Test;

import com.netflix.conductor.common.metadata.tasks.TaskDef;
import com.netflix.conductor.common.metadata.tasks.TaskType;
import com.netflix.conductor.common.metadata.workflow.WorkflowDef;
import com.netflix.conductor.core.config.ConductorProperties;
import com.netflix.conductor.core.execution.WorkflowExecutor;
import com.netflix.conductor.dao.QueueDAO;
import com.netflix.conductor.model.TaskModel;
import com.netflix.conductor.model.TaskModel.Status;
import com.netflix.conductor.model.WorkflowModel;

import static com.netflix.conductor.core.utils.Utils.DECIDER_QUEUE;

import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class TestWorkflowSweeper {

    private ConductorProperties properties;
    private WorkflowExecutor workflowExecutor;
    private WorkflowRepairService workflowRepairService;
    private QueueDAO queueDAO;
    private WorkflowSweeper workflowSweeper;

    private int defaultPostPoneOffSetSeconds = 1800;

    @Before
    public void setUp() {
        properties = mock(ConductorProperties.class);
        workflowExecutor = mock(WorkflowExecutor.class);
        queueDAO = mock(QueueDAO.class);
        workflowRepairService = mock(WorkflowRepairService.class);
        workflowSweeper =
                new WorkflowSweeper(
                        workflowExecutor, Optional.of(workflowRepairService), properties, queueDAO);
    }

    @Test
    public void testPostponeDurationForHumanTaskType() {
        WorkflowModel workflowModel = new WorkflowModel();
        workflowModel.setWorkflowId("1");
        TaskModel taskModel = new TaskModel();
        taskModel.setTaskId("task1");
        taskModel.setTaskType(TaskType.TASK_TYPE_HUMAN);
        taskModel.setStatus(Status.IN_PROGRESS);
        workflowModel.setTasks(List.of(taskModel));
        when(properties.getWorkflowOffsetTimeout())
                .thenReturn(Duration.ofSeconds(defaultPostPoneOffSetSeconds));
        workflowSweeper.unack(workflowModel, defaultPostPoneOffSetSeconds);
        verify(queueDAO)
                .setUnackTimeout(
                        DECIDER_QUEUE,
                        workflowModel.getWorkflowId(),
                        defaultPostPoneOffSetSeconds * 1000);
    }

    @Test
    public void testPostponeDurationForWaitTaskType() {
        WorkflowModel workflowModel = new WorkflowModel();
        workflowModel.setWorkflowId("1");
        TaskModel taskModel = new TaskModel();
        taskModel.setTaskId("task1");
        taskModel.setTaskType(TaskType.TASK_TYPE_WAIT);
        taskModel.setStatus(Status.IN_PROGRESS);
        workflowModel.setTasks(List.of(taskModel));
        when(properties.getWorkflowOffsetTimeout())
                .thenReturn(Duration.ofSeconds(defaultPostPoneOffSetSeconds));
        workflowSweeper.unack(workflowModel, defaultPostPoneOffSetSeconds);
        verify(queueDAO)
                .setUnackTimeout(
                        DECIDER_QUEUE,
                        workflowModel.getWorkflowId(),
                        defaultPostPoneOffSetSeconds * 1000);
    }

    @Test
    public void testPostponeDurationForWaitTaskTypeWithLongWaitTime() {
        long waitTimeout = 65845;
        WorkflowModel workflowModel = new WorkflowModel();
        workflowModel.setWorkflowId("1");
        TaskModel taskModel = new TaskModel();
        taskModel.setTaskId("task1");
        taskModel.setTaskType(TaskType.TASK_TYPE_WAIT);
        taskModel.setStatus(Status.IN_PROGRESS);
        taskModel.setWaitTimeout(System.currentTimeMillis() + waitTimeout);
        workflowModel.setTasks(List.of(taskModel));
        when(properties.getWorkflowOffsetTimeout())
                .thenReturn(Duration.ofSeconds(defaultPostPoneOffSetSeconds));
        workflowSweeper.unack(workflowModel, defaultPostPoneOffSetSeconds);
        verify(queueDAO)
                .setUnackTimeout(
                        DECIDER_QUEUE,
                        workflowModel.getWorkflowId(),
                        (waitTimeout / 1000L + 1L) * 1000L);
    }

    @Test
    public void testPostponeDurationForWaitTaskTypeWithWaitTime() {
        long waitTimeout = 180;
        WorkflowModel workflowModel = new WorkflowModel();
        workflowModel.setWorkflowId("1");
        TaskModel taskModel = new TaskModel();
        taskModel.setTaskId("task1");
        taskModel.setTaskType(TaskType.TASK_TYPE_WAIT);
        taskModel.setStatus(Status.IN_PROGRESS);
        taskModel.setWaitTimeout(System.currentTimeMillis() + waitTimeout);
        workflowModel.setTasks(List.of(taskModel));
        when(properties.getWorkflowOffsetTimeout())
                .thenReturn(Duration.ofSeconds(defaultPostPoneOffSetSeconds));
        workflowSweeper.unack(workflowModel, defaultPostPoneOffSetSeconds);
        verify(queueDAO)
                .setUnackTimeout(
                        DECIDER_QUEUE,
                        workflowModel.getWorkflowId(),
                        (waitTimeout / 1000 + 1) * 1000);
    }

    @Test
    public void testPostponeDurationForTaskInProgress() {
        WorkflowModel workflowModel = new WorkflowModel();
        workflowModel.setWorkflowId("1");
        TaskModel taskModel = new TaskModel();
        taskModel.setTaskId("task1");
        taskModel.setTaskType(TaskType.TASK_TYPE_SIMPLE);
        taskModel.setStatus(Status.IN_PROGRESS);
        workflowModel.setTasks(List.of(taskModel));
        when(properties.getWorkflowOffsetTimeout())
                .thenReturn(Duration.ofSeconds(defaultPostPoneOffSetSeconds));
        workflowSweeper.unack(workflowModel, defaultPostPoneOffSetSeconds);
        verify(queueDAO)
                .setUnackTimeout(
                        DECIDER_QUEUE,
                        workflowModel.getWorkflowId(),
                        defaultPostPoneOffSetSeconds * 1000);
    }

    @Test
    public void testPostponeDurationForTaskInProgressWithResponseTimeoutSet() {
        long responseTimeout = 200;
        WorkflowModel workflowModel = new WorkflowModel();
        workflowModel.setWorkflowId("1");
        TaskModel taskModel = new TaskModel();
        taskModel.setTaskId("task1");
        taskModel.setTaskType(TaskType.TASK_TYPE_SIMPLE);
        taskModel.setStatus(Status.IN_PROGRESS);
        taskModel.setResponseTimeoutSeconds(responseTimeout);
        workflowModel.setTasks(List.of(taskModel));
        when(properties.getWorkflowOffsetTimeout())
                .thenReturn(Duration.ofSeconds(defaultPostPoneOffSetSeconds));
        workflowSweeper.unack(workflowModel, defaultPostPoneOffSetSeconds);
        verify(queueDAO)
                .setUnackTimeout(
                        DECIDER_QUEUE, workflowModel.getWorkflowId(), (responseTimeout + 1) * 1000);
    }

    @Test
    public void testPostponeDurationForTaskInScheduled() {
        WorkflowModel workflowModel = new WorkflowModel();
        workflowModel.setWorkflowId("1");
        WorkflowDef workflowDef = new WorkflowDef();
        workflowModel.setWorkflowDefinition(workflowDef);
        TaskModel taskModel = new TaskModel();
        taskModel.setTaskId("task1");
        taskModel.setTaskType(TaskType.TASK_TYPE_SIMPLE);
        taskModel.setStatus(Status.SCHEDULED);
        taskModel.setReferenceTaskName("task1");
        workflowModel.setTasks(List.of(taskModel));
        when(properties.getWorkflowOffsetTimeout())
                .thenReturn(Duration.ofSeconds(defaultPostPoneOffSetSeconds));
        workflowSweeper.unack(workflowModel, defaultPostPoneOffSetSeconds);
        verify(queueDAO)
                .setUnackTimeout(
                        DECIDER_QUEUE,
                        workflowModel.getWorkflowId(),
                        defaultPostPoneOffSetSeconds * 1000);
    }

    @Test
    public void testPostponeDurationForTaskInScheduledWithWorkflowTimeoutSet() {
        long workflowTimeout = 1800;
        WorkflowModel workflowModel = new WorkflowModel();
        workflowModel.setWorkflowId("1");
        WorkflowDef workflowDef = new WorkflowDef();
        workflowDef.setTimeoutSeconds(workflowTimeout);
        workflowModel.setWorkflowDefinition(workflowDef);
        TaskModel taskModel = new TaskModel();
        taskModel.setTaskId("task1");
        taskModel.setTaskType(TaskType.TASK_TYPE_SIMPLE);
        taskModel.setStatus(Status.SCHEDULED);
        workflowModel.setTasks(List.of(taskModel));
        when(properties.getWorkflowOffsetTimeout())
                .thenReturn(Duration.ofSeconds(defaultPostPoneOffSetSeconds));
        workflowSweeper.unack(workflowModel, defaultPostPoneOffSetSeconds);
        verify(queueDAO)
                .setUnackTimeout(
                        DECIDER_QUEUE, workflowModel.getWorkflowId(), (workflowTimeout + 1) * 1000);
    }

    @Test
    public void testPostponeDurationForTaskInScheduledWithWorkflowTimeoutSetAndNoPollTimeout() {
        long workflowTimeout = 1800;
        WorkflowModel workflowModel = new WorkflowModel();
        workflowModel.setWorkflowId("1");
        WorkflowDef workflowDef = new WorkflowDef();
        workflowDef.setTimeoutSeconds(workflowTimeout);
        workflowModel.setWorkflowDefinition(workflowDef);
        TaskDef taskDef = new TaskDef();
        TaskModel taskModel = mock(TaskModel.class);
        workflowModel.setTasks(List.of(taskModel));
        when(taskModel.getTaskDefinition()).thenReturn(Optional.of(taskDef));
        when(taskModel.getStatus()).thenReturn(Status.SCHEDULED);
        when(properties.getWorkflowOffsetTimeout())
                .thenReturn(Duration.ofSeconds(defaultPostPoneOffSetSeconds));
        workflowSweeper.unack(workflowModel, defaultPostPoneOffSetSeconds);
        verify(queueDAO)
                .setUnackTimeout(
                        DECIDER_QUEUE, workflowModel.getWorkflowId(), (workflowTimeout + 1) * 1000);
    }

    @Test
    public void testPostponeDurationForTaskInScheduledWithNoWorkflowTimeoutSetAndNoPollTimeout() {
        WorkflowModel workflowModel = new WorkflowModel();
        workflowModel.setWorkflowId("1");
        WorkflowDef workflowDef = new WorkflowDef();
        workflowModel.setWorkflowDefinition(workflowDef);
        TaskDef taskDef = new TaskDef();
        TaskModel taskModel = mock(TaskModel.class);
        workflowModel.setTasks(List.of(taskModel));
        when(taskModel.getTaskDefinition()).thenReturn(Optional.of(taskDef));
        when(taskModel.getStatus()).thenReturn(Status.SCHEDULED);
        when(properties.getWorkflowOffsetTimeout())
                .thenReturn(Duration.ofSeconds(defaultPostPoneOffSetSeconds));
        workflowSweeper.unack(workflowModel, defaultPostPoneOffSetSeconds);
        verify(queueDAO)
                .setUnackTimeout(
                        DECIDER_QUEUE,
                        workflowModel.getWorkflowId(),
                        defaultPostPoneOffSetSeconds * 1000);
    }

    @Test
    public void testPostponeDurationForTaskInScheduledWithNoPollTimeoutSet() {
        WorkflowModel workflowModel = new WorkflowModel();
        workflowModel.setWorkflowId("1");
        TaskDef taskDef = new TaskDef();
        WorkflowDef workflowDef = new WorkflowDef();
        workflowModel.setWorkflowDefinition(workflowDef);
        TaskModel taskModel = mock(TaskModel.class);
        workflowModel.setTasks(List.of(taskModel));
        when(taskModel.getStatus()).thenReturn(Status.SCHEDULED);
        when(taskModel.getTaskDefinition()).thenReturn(Optional.of(taskDef));
        when(properties.getWorkflowOffsetTimeout())
                .thenReturn(Duration.ofSeconds(defaultPostPoneOffSetSeconds));
        workflowSweeper.unack(workflowModel, defaultPostPoneOffSetSeconds);
        verify(queueDAO)
                .setUnackTimeout(
                        DECIDER_QUEUE,
                        workflowModel.getWorkflowId(),
                        defaultPostPoneOffSetSeconds * 1000);
    }

    @Test
    public void testPostponeDurationForTaskInScheduledWithPollTimeoutSet() {
        int pollTimeout = 200;
        WorkflowModel workflowModel = new WorkflowModel();
        workflowModel.setWorkflowId("1");
        TaskDef taskDef = new TaskDef();
        taskDef.setPollTimeoutSeconds(pollTimeout);
        TaskModel taskModel = mock(TaskModel.class);
        ;
        workflowModel.setTasks(List.of(taskModel));
        when(taskModel.getStatus()).thenReturn(Status.SCHEDULED);
        when(taskModel.getTaskDefinition()).thenReturn(Optional.of(taskDef));
        when(properties.getWorkflowOffsetTimeout())
                .thenReturn(Duration.ofSeconds(defaultPostPoneOffSetSeconds));
        workflowSweeper.unack(workflowModel, defaultPostPoneOffSetSeconds);
        verify(queueDAO)
                .setUnackTimeout(
                        DECIDER_QUEUE, workflowModel.getWorkflowId(), (pollTimeout + 1) * 1000);
    }

    @Test
    public void testWorkflowOffsetJitter() {
        long offset = 45;
        for (int i = 0; i < 10; i++) {
            long offsetWithJitter = workflowSweeper.workflowOffsetWithJitter(offset);
            assertTrue(offsetWithJitter >= 30);
            assertTrue(offsetWithJitter <= 60);
        }
    }
}
```

Reply:
I also updated the branch in the same way.

Reply:
@manan164 Thanks for the change you suggest, you are right!

Reply:
@manan164  I updated the branch with the changes and tests.